domain/externalcontroller/state/state.go

```diff
@@ -185,33 +168,92 @@ func (st *State) updateExternalControllerTx(
 		Alias:  sql.NullString{String: ci.Alias, Valid: true},
 		CACert: ci.CACert,
 	}
-	if err := tx.Query(ctx, stmts.upsertController, externalController).Run(); err != nil {
+
+	upsertControllerQuery := `
+INSERT INTO external_controller (uuid, alias, ca_cert)
+VALUES ($Controller.*)
+  ON CONFLICT(uuid) DO UPDATE SET alias=excluded.alias, ca_cert=excluded.ca_cert
+`
+	upsertControllerStmt, err := st.Prepare(upsertControllerQuery, externalController)
+	if err != nil {
+		return errors.Annotatef(err, "preparing %q:", upsertControllerQuery)
+	}
+
+	if err := tx.Query(ctx, upsertControllerStmt, externalController).Run(); err != nil {
 		return errors.Trace(domain.CoerceError(err))
 	}
 
-	cIDs := uuids(ci.Addrs)
+	cIDs := ControllerUUIDs(ci.Addrs)
 	if len(cIDs) == 0 {
-		cIDs = uuids{}
+		cIDs = ControllerUUIDs{}
```

Comment:
**suggestion:** This kind of hack worth a comment. In golang, empty slice should behave as nil slice (that a convention).

I think that either:

* It never has been necessary (we should remove this check)
* It was necessary (when we uses the previous db) but it's not in sqlair (we should remove thise check)
* It is still necessary (maybe because nil slice are converted to null in `sqlair`, by design choice), but we should add a comment that this strange check is *intentional* and *why*

FYK: https://go.dev/play/p/keEpfAeByWu

Reply:
I'm not sure how this ended up here but it is entirely pointless. I've removed it.

Reply:
Probably because of the way the previous db lib works.

In some context (like DB, or even JSON encoding), NULL is a thing. So NULL array isn't an empty array.

In golang, nil slices should behave like empty one (purely conventional, but if we don't respect it, we end up with function which doesn't behave like builtin function and standard lib functionnalities). So when we reach an interface between golang and "something where NULL array/slice are actually a *thing*", we should be especially carefull, since it may cause very subtile bugs.

My two cents ;)